keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,763 @@
+      because the policy does not affect this since we are not replacing the pods.
+
+### ReplicaSet Status and Deployment Status Changes
+
+We should keep the current counting behavior for `.status.replicas` regardless of the
+PodReplacementPolicy, for backwards compatibility reasons. Current consumers of the Deployment API
+are only expecting non-terminating pods to be present in this field.
+
+To satisfy the requirement for tracking terminating pods, and for implementation purposes,
+we propose a new field `.status.terminatingReplicas` to the ReplicaSet's and Deployment's
+status.
+
+### API
+
+```golang
+// DeploymentPodReplacementPolicy specifies the policy for creating Deployment Pod replacements.
+// Default is a mixed behavior depending on the DeploymentStrategy
+// +enum 
+type DeploymentPodReplacementPolicy string
+const (
+// Terminating policy creates replacement Pods when the old Pods start
+// terminating (has a .metadata.deletionTimestamp). The total number of
+// Deployment Pods can be greater than specified by the Deployment's
+// .spec.replicas and the DeploymentStrategy.
+Terminating DeploymentPodReplacementPolicy = "Terminating"
+// Terminated policy creates replacement Pods only when the old Pods are
+// fully terminated (reach Succeeded or Failed phase). The old Pods are
+// subsequently removed. The total number of the Deployment Pods is
+// limited by the Deployment's .spec.replicas and the DeploymentStrategy.
+Terminated DeploymentPodReplacementPolicy = "Terminated"
+)
+```
+
+```golang
+type DeploymentSpec struct {
+    ...
+    // podReplacementPolicy specifies when to create replacement Pods. 
+	// Possible values are:
+    // - Terminating policy creates replacement Pods when the old Pods start
+	//   terminating (has a .metadata.deletionTimestamp). The total number of
+	//   Deployment Pods can be greater than specified by the Deployment's
+	//   .spec.replicas and the DeploymentStrategy.
+    // - Terminated policy creates replacement Pods only when the old Pods are
+	//   fully terminated (reach Succeeded or Failed phase). The old Pods are
+	//   subsequently removed. The total number of the Deployment Pods is
+	//   limited by the Deployment's .spec.replicas and the DeploymentStrategy.
+    //
+    // The default behavior when the policy is not specified depends on the DeploymentStrategy:
+	// - Recreate strategy uses Terminated behavior when recreating the deployment,
+	//   but uses Terminating when scaling the deployment.
+	// - RollingUpdate strategy uses Terminating behavior for both rolling out and
+	//   scaling the deployments.
+	//
+	// This is an alpha field. Enable DeploymentPodReplacementPolicy to be able to
+	// use this field.
+    // +optional
+    PodReplacementPolicy *DeploymentPodReplacementPolicy
+    ...
+}
+```
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+	// The number of terminating replicas for this replica set.
```

Comment:
As the number of replica status counters increases, I'd like to make sure the relationships between the counts are clear.  We have:

```go
	// Replicas is the number of actual replicas.
	Replicas int32

	// The number of pods that have labels matching the labels of the pod template of the replicaset.
	// +optional
	FullyLabeledReplicas int32

	// The number of ready replicas for this replica set.
	// +optional
	ReadyReplicas int32

	// The number of available replicas (ready for at least minReadySeconds) for this replica set.
	// +optional
	AvailableReplicas int32
```

Does `FullyLabeledReplicas` , `Replicas` or `ReadyReplicas` count terminating replicas? 

Are we guaranteed `ReadyReplicas + TerminatingReplicas = Replicas` ?

Reply:
Same answer as https://github.com/kubernetes/enhancements/pull/4357#discussion_r1476709397. 

Strange thing, not sure why are the comments a bit different in the `/api/` and `/apis/` types?

Also the style of each field is a bit different. 

Can we go over the fields and make the documentation consistent when we do the PR in k/k?

Reply:
> Can we go over the fields and make the documentation consistent when we do the PR in k/k?

That would be great.